### ISSUE TITLE
🐛 BUG: Astro 0.24.x with UnCSS or PurgeCSS remove styles in components

### ISSUE THREAD

**Issue by djmtype** [withastro/astro#2890]
### What version of `astro` are you using?

0.24.3

### What package manager are you using?

npm

### What operating system are you using?

Mac

### Describe the Bug

Upon build, tree-shaking css with either postcss-uncss (https://www.npmjs.com/package/postcss-uncss) or PurgeCSS (https://www.npmjs.com/package/@fullhuman/postcss-purgecss) seems to remove styles within Astro components when using the <style> tag. 

This bug isn't always persistent either. Sometimes it builds correctly. So, try building and previewing the build a few times. 

Try searching for `.banner` in the compiled css file as an indicator. This class should exist.

There was no issue with Astro v0.23.x, and everything seems fine with Astro v0.24.x only with the `--legacy-build` flag. 

Using Node v16.13.0, MacOS 12.3 on m1 ARM.

I can not build the project on stackblitz because I don't have a pro account, but I've tested this locally. 

Run `npm run build` then `npm run preview`. When running the preview of the build, the Astro Starter index page should visually appear off due to the bug as shown in the attachment.

![Screen Shot 2022-03-25 at 6 23 43 PM](https://user-images.githubusercontent.com/930713/160210096-a8ec3c6d-6c21-44f2-9b69-4fe6004c25e2.png)

### Link to Minimal Reproducible Example

https://stackblitz.com/edit/github-sndlkb

### Participation

- [ ] I am willing to submit a pull request for this issue.

**Comment by natemoo-re** [comment]
I know `uncss` and `unocss` are not the same, but is this a possible duplicate of https://github.com/withastro/astro/issues/1971 ? I suspect the fix might be the same.

**Comment by djmtype** [comment]
> I know `uncss` and `unocss` are not the same, but is this a possible duplicate of https://github.com/withastro/astro/issues/1971 ? I suspect the fix might be the same.

I'm not sure, and I should also test this during dev, not just build to see if there are any differences.

**Comment by djmtype** [comment]
> I know `uncss` and `unocss` are not the same, but is this a possible duplicate of #1971 ? I suspect the fix might be the same.

@natemoo-re I'm not entirely sure as the issue with uncss is specifically with styles within components only and only during `build`. I tested uncss during `dev` and it works just fine. 

Like I said, sometimes during build, it does include (or doesn't remove) the necessary styles. It's completely random – at least to me. 

Here's also a diff checker of the Astro Starter. 
https://www.diffchecker.com/RNr38iiy

The red highlighted lines are the styles being removed by uncss when the build isn't succeeding properly.

**Comment by djmtype** [comment]
In regards to PurgeCSS, using `/* purgecss start ignore */` … `/* purgecss end ignore */` within the component's `<style>` tag has no effect. 

I'd have to specify a safelist rule in postcss.config.js,  `deep: [/banner$/]` so it doesn't remove `.banner` from the css for example.

**Comment by djmtype** [comment]
In regards to styles within the <style> tag, any style within :global is being processed properly so long as those style are within `/* purgecss start ignore */` … `/* purgecss end ignore */`.

This remains to be an issue in Astro 1.0.0-beta10

The following `.banner` and `.text` class names are being used. 

In this example, `:global(.banner)` and `.text` will be removed regardless. Therefore, processed incorrectly.

```
// component.astro

<style>
:global(.banner) {
background-color: blue;
}
.text {
color: yellow;
}
</style>
```

In this example, `:global(.banner)` and `.text` is are wrapped with purgecss' ignore statement. Any `:global` style will be processed correctly, while all other styles will be removed.

```
// component.astro

<style>
/* purgecss start ignore */

:global(.banner) {
background-color: blue;
}
.text {
color: yellow;
}
/* purgecss end ignore */
</style>
```

**Comment by djmtype** [comment]
I don't know if this adds any value to this issue, but I'm also using Vue components in this Astro project and they get processed correctly by purgecss, and I didn't have to wrap them inside a purgecss ignore comment either.

**Comment by FredKSchott** [comment]
I can't reproduce in `v1.0.0-beta`, hoping that this was fixed in latest. 

If you can still reproduce after upgrading to the latest version of Astro and all other dependencies, we'll need a reliable reproduction before we can debug. Please file a new issue if you can still reproduce!
